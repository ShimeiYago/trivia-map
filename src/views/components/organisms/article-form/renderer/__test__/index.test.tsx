import { shallow, ShallowWrapper } from 'enzyme';
import { Renderer, Props } from '..';

let wrapper: ShallowWrapper<Props, unknown, Renderer>;

const basicProps: Props = {
  title: '',
  description: '',
  imageDataUrl: null,
  position: {
    lat: 0,
    lng: 0,
    park: 'S',
  },
  submittingState: 'waiting',
  fetchingState: 'waiting',
  isFormEditting: false,

  updateFormField: jest.fn(),
  submitNewArticle: jest.fn(),
  submitEdittedArticle: jest.fn(),
  fetchArticle: jest.fn(),
  initialize: jest.fn(),
  updateIsEditting: jest.fn(),
};

describe('Shallow Snapshot Tests', () => {
  beforeEach(() => {
    wrapper = shallow(<Renderer {...basicProps} />);
  });

  it('basic', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('position is not selected yet', () => {
    wrapper.setProps({ position: undefined });
    expect(wrapper).toMatchSnapshot();
  });

  it('with validation error', () => {
    wrapper.setProps({
      formError: {
        errorTitle: 'Inputted contents have errors.',
        fieldErrors: { position: 'position is not selected' },
      },
    });
    expect(wrapper).toMatchSnapshot();
  });

  it('loading', () => {
    wrapper.setProps({
      fetchingState: 'loading',
    });
    expect(wrapper).toMatchSnapshot();
  });

  it('with close button', () => {
    wrapper.setProps({
      onClose: jest.fn(),
    });
    expect(wrapper).toMatchSnapshot();
  });

  it('with image', () => {
    wrapper.setProps({
      imageDataUrl: 'https://image-data.jpg',
    });
    expect(wrapper).toMatchSnapshot();
  });
});

describe('constructor', () => {
  it('should not call initialize prop with isEditting mode', () => {
    wrapper = shallow(<Renderer {...basicProps} isFormEditting />);
    expect(basicProps.initialize).not.toBeCalled();
  });
});

describe('componentDidMount', () => {
  it('fetchArticle if postId exists', () => {
    wrapper = shallow(<Renderer {...basicProps} />);
    wrapper.setProps({ postId: 100 });
    const instance = wrapper.instance();
    instance.componentDidMount();

    expect(basicProps.fetchArticle).toBeCalled();
  });
});

describe('componentDidUpdate', () => {
  beforeEach(() => {
    wrapper = shallow(<Renderer {...basicProps} />);
  });

  it('should call scrollIntoView', () => {
    wrapper.setProps({ submittingState: 'error' });
    const instance = wrapper.instance();

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const headerRef: React.RefObject<any> = {
      current: {
        scrollIntoView: jest.fn(),
      },
    };
    instance.headerRef = headerRef;

    instance.componentDidUpdate(basicProps);
    expect(headerRef.current.scrollIntoView).toBeCalled();
  });

  it('should call fetchArticle if postId exists', () => {
    wrapper.setProps({ postId: 100 });
    const instance = wrapper.instance();
    instance.componentDidUpdate(basicProps);

    expect(basicProps.fetchArticle).toBeCalled();
  });
});

describe('componentWillUnmount', () => {
  beforeEach(() => {
    wrapper = shallow(<Renderer {...basicProps} />);
  });

  it('should not call initialize if isFormEditting', () => {
    wrapper.setProps({ isFormEditting: true });
    const instance = wrapper.instance();
    instance.componentWillUnmount();
    expect(instance.props.initialize).toBeCalledTimes(1);
  });

  it('should call initialize if not isFormEditting', () => {
    const instance = wrapper.instance();
    instance.componentWillUnmount();
    expect(instance.props.initialize).toBeCalledTimes(2);
  });
});

describe('handleChangeTitle', () => {
  it('call updateTitle', () => {
    wrapper = shallow(<Renderer {...basicProps} />);
    const instance = wrapper.instance();

    const changeTitleEvent = { target: { value: 'new title' } };
    instance['handleChangeTitle'](
      changeTitleEvent as React.ChangeEvent<HTMLInputElement>,
    );

    expect(basicProps.updateFormField).toBeCalled();
  });
});

describe('handleChangeDescription', () => {
  it('call updateContent', () => {
    wrapper = shallow(<Renderer {...basicProps} />);
    const instance = wrapper.instance();

    const changeContentEvent = { target: { value: 'new content' } };
    instance['handleChangeDescription'](
      changeContentEvent as React.ChangeEvent<HTMLInputElement>,
    );

    expect(basicProps.updateFormField).toBeCalled();
  });
});

describe('handleSubmitButton', () => {
  beforeEach(() => {
    wrapper = shallow(<Renderer {...basicProps} />);
  });

  it('call submitNewArticle if postId does not exists', () => {
    const instance = wrapper.instance();
    instance['handleSubmitButton']();

    expect(basicProps.submitNewArticle).toBeCalled();
  });

  it('call submitEdittedArticle if postId exists', () => {
    wrapper.setProps({ postId: 100 });
    const instance = wrapper.instance();

    instance['handleSubmitButton']();

    expect(basicProps.submitEdittedArticle).toBeCalled();
  });
});

describe('handleFileInputChange', () => {
  beforeEach(() => {
    wrapper = shallow(<Renderer {...basicProps} />);
  });

  it('set imageData when image is uploaded', async () => {
    const blob = new Blob(['image-data']);
    const file = new File([blob], 'test.jpeg', {
      type: 'image/jpeg',
    });
    const event = {
      target: {
        files: [file],
      },
    } as unknown as React.ChangeEvent<HTMLInputElement>;
    const instance = wrapper.instance();
    await instance['handleFileInputChange'](event);

    expect(instance.props.updateFormField).not.toBeCalledWith({
      imageDataUrl: null,
    });
  });

  it('do not set imageData with no image', async () => {
    const event = {
      target: {
        files: [],
      },
    } as unknown as React.ChangeEvent<HTMLInputElement>;
    const instance = wrapper.instance();
    await instance['handleFileInputChange'](event);

    expect(instance.props.updateFormField).toBeCalledWith({
      imageDataUrl: null,
    });
  });
});

describe('handleDeleteImage', () => {
  beforeEach(() => {
    wrapper = shallow(<Renderer {...basicProps} />);
  });

  it('should set null as imageDataUrl', () => {
    const instance = wrapper.instance();
    instance['handleDeleteImage']();

    expect(instance.props.updateFormField).toBeCalledWith({
      imageDataUrl: null,
    });
  });
});
