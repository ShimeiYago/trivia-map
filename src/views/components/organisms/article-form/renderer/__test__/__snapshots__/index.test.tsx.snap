// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Shallow Snapshot Tests basic 1`] = `
<Fragment>
  <ForwardRef(Container)
    maxWidth="sm"
    sx={
      Object {
        "overflow": "scroll",
        "p": 3,
      }
    }
  >
    <ForwardRef(Stack)
      spacing={3}
    >
      <ForwardRef(Typography)
        align="center"
        component="h2"
        variant="h4"
      >
        新しいトリビアを追加
      </ForwardRef(Typography)>
      <ForwardRef(TextField)
        disabled={false}
        error={false}
        label="タイトル"
        onChange={[Function]}
        value=""
        variant="standard"
      />
      <ImageField
        onChange={[Function]}
      />
      <ForwardRef(TextField)
        disabled={false}
        error={false}
        label="説明文"
        multiline={true}
        onChange={[Function]}
        rows={4}
        value=""
      />
      <BoxField
        disabled={false}
        status="valid"
      >
        <ForwardRef(Box)
          sx={
            Object {
              "position": "relative",
            }
          }
        >
          <TriviaMap
            disabled={true}
            doNotShowPostMarkers={true}
            height={300}
            initCenter={
              Object {
                "lat": 0,
                "lng": 0,
                "park": "S",
              }
            }
            initZoom={3}
            shouldCurrentPositionAsyncWithForm={true}
          />
          <ForwardRef(Box)
            sx={
              Object {
                "backgroundColor": "rgba(0,0,0,0.5)",
                "height": "100%",
                "left": 0,
                "position": "absolute",
                "top": 0,
                "width": "100%",
                "zIndex": 2000,
              }
            }
          />
          <ForwardRef(Box)
            sx={
              Object {
                "fontWeight": "bold",
                "left": "50%",
                "position": "absolute",
                "top": "70%",
                "transform": "translate(-50%,-50%)",
                "width": "90%",
                "zIndex": 2000,
              }
            }
          >
            <ForwardRef(Typography)
              color="white"
              component="div"
              fontSize="14"
              textAlign="center"
              variant="inherit"
            >
              この位置が選択されています。
              <br />
              位置を選び直す場合はここをタップしてください。
            </ForwardRef(Typography)>
          </ForwardRef(Box)>
        </ForwardRef(Box)>
      </BoxField>
      <ForwardRef(LoadingButton)
        disabled={false}
        loading={false}
        onClick={[Function]}
        variant="outlined"
      >
        保存して公開
      </ForwardRef(LoadingButton)>
    </ForwardRef(Stack)>
  </ForwardRef(Container)>
</Fragment>
`;

exports[`Shallow Snapshot Tests loading 1`] = `
<Fragment>
  <ForwardRef(Container)
    maxWidth="sm"
    sx={
      Object {
        "overflow": "scroll",
        "p": 3,
      }
    }
  >
    <ForwardRef(Stack)
      spacing={3}
    >
      <ForwardRef(Typography)
        align="center"
        component="h2"
        variant="h4"
      >
        新しいトリビアを追加
      </ForwardRef(Typography)>
      <ForwardRef(Box)
        sx={
          Object {
            "display": "flex",
            "justifydescription": "center",
          }
        }
      >
        <ForwardRef(CircularProgress) />
      </ForwardRef(Box)>
      <ForwardRef(TextField)
        disabled={true}
        error={false}
        label="タイトル"
        onChange={[Function]}
        value=""
        variant="standard"
      />
      <ImageField
        onChange={[Function]}
      />
      <ForwardRef(TextField)
        disabled={true}
        error={false}
        label="説明文"
        multiline={true}
        onChange={[Function]}
        rows={4}
        value=""
      />
      <BoxField
        disabled={true}
        status="valid"
      >
        <ForwardRef(Box)
          sx={
            Object {
              "position": "relative",
            }
          }
        >
          <TriviaMap
            disabled={true}
            doNotShowPostMarkers={true}
            height={300}
            initCenter={
              Object {
                "lat": 0,
                "lng": 0,
                "park": "S",
              }
            }
            initZoom={3}
            shouldCurrentPositionAsyncWithForm={true}
          />
          <ForwardRef(Box)
            sx={
              Object {
                "backgroundColor": "rgba(0,0,0,0.5)",
                "height": "100%",
                "left": 0,
                "position": "absolute",
                "top": 0,
                "width": "100%",
                "zIndex": 2000,
              }
            }
          />
          <ForwardRef(Box)
            sx={
              Object {
                "fontWeight": "bold",
                "left": "50%",
                "position": "absolute",
                "top": "70%",
                "transform": "translate(-50%,-50%)",
                "width": "90%",
                "zIndex": 2000,
              }
            }
          />
        </ForwardRef(Box)>
      </BoxField>
      <ForwardRef(LoadingButton)
        disabled={true}
        loading={false}
        onClick={[Function]}
        variant="outlined"
      >
        保存して公開
      </ForwardRef(LoadingButton)>
    </ForwardRef(Stack)>
  </ForwardRef(Container)>
</Fragment>
`;

exports[`Shallow Snapshot Tests position is not selected yet 1`] = `
<Fragment>
  <ForwardRef(Container)
    maxWidth="sm"
    sx={
      Object {
        "overflow": "scroll",
        "p": 3,
      }
    }
  >
    <ForwardRef(Stack)
      spacing={3}
    >
      <ForwardRef(Typography)
        align="center"
        component="h2"
        variant="h4"
      >
        新しいトリビアを追加
      </ForwardRef(Typography)>
      <ForwardRef(TextField)
        disabled={false}
        error={false}
        label="タイトル"
        onChange={[Function]}
        value=""
        variant="standard"
      />
      <ImageField
        onChange={[Function]}
      />
      <ForwardRef(TextField)
        disabled={false}
        error={false}
        label="説明文"
        multiline={true}
        onChange={[Function]}
        rows={4}
        value=""
      />
      <BoxField
        disabled={false}
        status="normal"
      >
        <ForwardRef(Box)
          sx={
            Object {
              "position": "relative",
            }
          }
        >
          <TriviaMap
            disabled={true}
            doNotShowPostMarkers={true}
            height={300}
            initZoom={3}
            shouldCurrentPositionAsyncWithForm={true}
          />
          <ForwardRef(Box)
            sx={
              Object {
                "backgroundColor": "rgba(0,0,0,0.5)",
                "height": "100%",
                "left": 0,
                "position": "absolute",
                "top": 0,
                "width": "100%",
                "zIndex": 2000,
              }
            }
          />
          <ForwardRef(Box)
            sx={
              Object {
                "fontWeight": "bold",
                "left": "50%",
                "position": "absolute",
                "top": "70%",
                "transform": "translate(-50%,-50%)",
                "width": "90%",
                "zIndex": 2000,
              }
            }
          >
            <ForwardRef(Typography)
              color="white"
              component="div"
              fontSize="14"
              textAlign="center"
              variant="inherit"
            >
              位置が選択されていません。
              <br />
              ここをタップして位置を選択してください。
            </ForwardRef(Typography)>
          </ForwardRef(Box)>
        </ForwardRef(Box)>
      </BoxField>
      <ForwardRef(LoadingButton)
        disabled={false}
        loading={false}
        onClick={[Function]}
        variant="outlined"
      >
        保存して公開
      </ForwardRef(LoadingButton)>
    </ForwardRef(Stack)>
  </ForwardRef(Container)>
</Fragment>
`;

exports[`Shallow Snapshot Tests with close button 1`] = `
<Fragment>
  <ForwardRef(Box)
    sx={
      Object {
        "backgroundColor": "white",
        "display": "block",
        "pl": 1,
        "pt": 1,
        "width": "100%",
      }
    }
  >
    <CloseFormButton
      onClose={[MockFunction]}
    />
  </ForwardRef(Box)>
  <ForwardRef(Container)
    maxWidth="sm"
    sx={
      Object {
        "overflow": "scroll",
        "p": 3,
      }
    }
  >
    <ForwardRef(Stack)
      spacing={3}
    >
      <ForwardRef(Typography)
        align="center"
        component="h2"
        variant="h4"
      >
        新しいトリビアを追加
      </ForwardRef(Typography)>
      <ForwardRef(TextField)
        disabled={false}
        error={false}
        label="タイトル"
        onChange={[Function]}
        value=""
        variant="standard"
      />
      <ImageField
        onChange={[Function]}
      />
      <ForwardRef(TextField)
        disabled={false}
        error={false}
        label="説明文"
        multiline={true}
        onChange={[Function]}
        rows={4}
        value=""
      />
      <BoxField
        disabled={false}
        status="valid"
      >
        <ForwardRef(Box)
          sx={
            Object {
              "position": "relative",
            }
          }
        >
          <TriviaMap
            disabled={true}
            doNotShowPostMarkers={true}
            height={300}
            initCenter={
              Object {
                "lat": 0,
                "lng": 0,
                "park": "S",
              }
            }
            initZoom={3}
            shouldCurrentPositionAsyncWithForm={true}
          />
          <ForwardRef(Box)
            sx={
              Object {
                "backgroundColor": "rgba(0,0,0,0.5)",
                "height": "100%",
                "left": 0,
                "position": "absolute",
                "top": 0,
                "width": "100%",
                "zIndex": 2000,
              }
            }
          />
          <ForwardRef(Box)
            sx={
              Object {
                "fontWeight": "bold",
                "left": "50%",
                "position": "absolute",
                "top": "70%",
                "transform": "translate(-50%,-50%)",
                "width": "90%",
                "zIndex": 2000,
              }
            }
          >
            <ForwardRef(Typography)
              color="white"
              component="div"
              fontSize="14"
              textAlign="center"
              variant="inherit"
            >
              この位置が選択されています。
              <br />
              位置を選び直す場合はここをタップしてください。
            </ForwardRef(Typography)>
          </ForwardRef(Box)>
        </ForwardRef(Box)>
      </BoxField>
      <ForwardRef(LoadingButton)
        disabled={false}
        loading={false}
        onClick={[Function]}
        variant="outlined"
      >
        保存して公開
      </ForwardRef(LoadingButton)>
    </ForwardRef(Stack)>
  </ForwardRef(Container)>
</Fragment>
`;

exports[`Shallow Snapshot Tests with image 1`] = `
<Fragment>
  <ForwardRef(Container)
    maxWidth="sm"
    sx={
      Object {
        "overflow": "scroll",
        "p": 3,
      }
    }
  >
    <ForwardRef(Stack)
      spacing={3}
    >
      <ForwardRef(Typography)
        align="center"
        component="h2"
        variant="h4"
      >
        新しいトリビアを追加
      </ForwardRef(Typography)>
      <ForwardRef(TextField)
        disabled={false}
        error={false}
        label="タイトル"
        onChange={[Function]}
        value=""
        variant="standard"
      />
      <DeletableImage
        borderRadius={true}
        height="200px"
        objectFit="cover"
        onDelete={[Function]}
        src="https://image-data.jpg"
        width="full"
      />
      <ForwardRef(TextField)
        disabled={false}
        error={false}
        label="説明文"
        multiline={true}
        onChange={[Function]}
        rows={4}
        value=""
      />
      <BoxField
        disabled={false}
        status="valid"
      >
        <ForwardRef(Box)
          sx={
            Object {
              "position": "relative",
            }
          }
        >
          <TriviaMap
            disabled={true}
            doNotShowPostMarkers={true}
            height={300}
            initCenter={
              Object {
                "lat": 0,
                "lng": 0,
                "park": "S",
              }
            }
            initZoom={3}
            shouldCurrentPositionAsyncWithForm={true}
          />
          <ForwardRef(Box)
            sx={
              Object {
                "backgroundColor": "rgba(0,0,0,0.5)",
                "height": "100%",
                "left": 0,
                "position": "absolute",
                "top": 0,
                "width": "100%",
                "zIndex": 2000,
              }
            }
          />
          <ForwardRef(Box)
            sx={
              Object {
                "fontWeight": "bold",
                "left": "50%",
                "position": "absolute",
                "top": "70%",
                "transform": "translate(-50%,-50%)",
                "width": "90%",
                "zIndex": 2000,
              }
            }
          >
            <ForwardRef(Typography)
              color="white"
              component="div"
              fontSize="14"
              textAlign="center"
              variant="inherit"
            >
              この位置が選択されています。
              <br />
              位置を選び直す場合はここをタップしてください。
            </ForwardRef(Typography)>
          </ForwardRef(Box)>
        </ForwardRef(Box)>
      </BoxField>
      <ForwardRef(LoadingButton)
        disabled={false}
        loading={false}
        onClick={[Function]}
        variant="outlined"
      >
        保存して公開
      </ForwardRef(LoadingButton)>
    </ForwardRef(Stack)>
  </ForwardRef(Container)>
</Fragment>
`;

exports[`Shallow Snapshot Tests with validation error 1`] = `
<Fragment>
  <ForwardRef(Container)
    maxWidth="sm"
    sx={
      Object {
        "overflow": "scroll",
        "p": 3,
      }
    }
  >
    <ForwardRef(Stack)
      spacing={3}
    >
      <ForwardRef(Typography)
        align="center"
        component="h2"
        variant="h4"
      >
        新しいトリビアを追加
      </ForwardRef(Typography)>
      <HeaderErrorMessages
        errorTitle="Inputted contents have errors."
      />
      <ForwardRef(TextField)
        disabled={false}
        error={false}
        label="タイトル"
        onChange={[Function]}
        value=""
        variant="standard"
      />
      <ImageField
        onChange={[Function]}
      />
      <ForwardRef(TextField)
        disabled={false}
        error={false}
        label="説明文"
        multiline={true}
        onChange={[Function]}
        rows={4}
        value=""
      />
      <BoxField
        disabled={false}
        helperText="position is not selected"
        status="error"
      >
        <ForwardRef(Box)
          sx={
            Object {
              "position": "relative",
            }
          }
        >
          <TriviaMap
            disabled={true}
            doNotShowPostMarkers={true}
            height={300}
            initCenter={
              Object {
                "lat": 0,
                "lng": 0,
                "park": "S",
              }
            }
            initZoom={3}
            shouldCurrentPositionAsyncWithForm={true}
          />
          <ForwardRef(Box)
            sx={
              Object {
                "backgroundColor": "rgba(0,0,0,0.5)",
                "height": "100%",
                "left": 0,
                "position": "absolute",
                "top": 0,
                "width": "100%",
                "zIndex": 2000,
              }
            }
          />
          <ForwardRef(Box)
            sx={
              Object {
                "fontWeight": "bold",
                "left": "50%",
                "position": "absolute",
                "top": "70%",
                "transform": "translate(-50%,-50%)",
                "width": "90%",
                "zIndex": 2000,
              }
            }
          >
            <ForwardRef(Typography)
              color="white"
              component="div"
              fontSize="14"
              textAlign="center"
              variant="inherit"
            >
              この位置が選択されています。
              <br />
              位置を選び直す場合はここをタップしてください。
            </ForwardRef(Typography)>
          </ForwardRef(Box)>
        </ForwardRef(Box)>
      </BoxField>
      <ForwardRef(LoadingButton)
        disabled={false}
        loading={false}
        onClick={[Function]}
        variant="outlined"
      >
        保存して公開
      </ForwardRef(LoadingButton)>
    </ForwardRef(Stack)>
  </ForwardRef(Container)>
</Fragment>
`;
